### PR TITLE
Implement nats

### PR DESCRIPTION
**Pull request by nmtien**
Pull request for #19

### CODE REVIEW

File: nats/routez.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+type: json
+data:
+  url: /routez
+  server: ${NATS_HOSTPORT}
+
+observation:
+  - name: connz
```
Comment (sivasamyk) [inline nats/routez.yml]: the name should be changed to `routez`

File: nats/subsz.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        source: num_cache
+        type: long_gauge
+        label: cache
+        description: The number of result sets in the cache
+
+      - name: cache.hit.rate
+        source: cache_hit_rate
+        type: double_gauge
+        label: cache hit rate
+        description: The rate matches are being retrieved from cache
```
Comment (sivasamyk) [inline nats/subsz.yml]: what is the unit for this rate metrics - `%` or `hits/sec`?  Can we get the actual hit count & num requests as a metric?

File: nats/vaz.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        source: in_msgs
+        type: long_gauge
+        label: incoming messages
+        description: The amount of incoming messages
+
+      - name: bytes.in
+        source: in_bytes
+        type: long_gauge
+        label: incoming bytes
+        description: The amount of incoming bytes
```
Comment (sivasamyk) [inline nats/vaz.yml]: the unit is missing. I assume it is `bytes`. Same for outgoing bytes

File: nats/subsz.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    path: $.
+
+    metric:
+      - name: subscriptions
+        source: num_subscriptions
+        type: long_gauge
+        label: subscriptions
+        description: The number of subscriptions
+
+      - name: inserts
```
Comment (bsmid) [inline nats/subsz.yml]: Maybe rename to `subscriptions.inserts`, to make it clear these are subscriptions related? Renaming label to `subscriptions list inserts` would maybe also be OK. And the same for other metrics below.

File: nats/subsz.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+      - name: removes
+        source: num_removes
+        type: long_gauge
+        label: removes
+        description: The number of remove operations in subscriptions list
+
+      - name: matches
+        source: num_matches
+        type: long_gauge
```
Comment (bsmid) [inline nats/subsz.yml]: Are you sure this is a gauge? It sounds like some kind of a cumulative.

File: nats/subsz.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        label: matches
+        description: The number of times a match is found for a subscription
+
+      - name: cache
```
Comment (bsmid) [inline nats/subsz.yml]: Rename to `cache.size`? If yes, also rename the label accordingly.

File: nats/vaz.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  url: /varz
+  server: ${NATS_HOSTPORT}
+
+observation:
+  - name: varz
+    metricNamespace: nats
+    path: $.
+
+    metric:
+      - name: msgs.in
```
Comment (bsmid) [inline nats/vaz.yml]: We could name it `messages.in`.

File: nats/vaz.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        label: incoming messages
+        description: The amount of incoming messages
+
+      - name: bytes.in
+        source: in_bytes
+        type: long_gauge
+        label: incoming bytes
+        description: The amount of incoming bytes
+
+      - name: msgs.out
```
Comment (bsmid) [inline nats/vaz.yml]: `messages.out`

File: nats/routez.yml (hunk 1 above)
Comment (nmtien) [reply nats/routez.yml]: Done

File: nats/subsz.yml (hunk 2 above)
Comment (nmtien) [reply nats/subsz.yml]: it's value range is [0-1], so it is %, in this case should i multiple value by 100, to change unit to '%'

File: nats/vaz.yml (hunk 3 above)
Comment (nmtien) [reply nats/vaz.yml]: Done

File: nats/subsz.yml (hunk 4 above)
Comment (nmtien) [reply nats/subsz.yml]: OK. DONE

File: nats/subsz.yml (hunk 6 above)
Comment (nmtien) [reply nats/subsz.yml]: OK, i change to cache.size

File: nats/vaz.yml (hunk 7 above)
Comment (nmtien) [reply nats/vaz.yml]: Ok

File: nats/vaz.yml (hunk 8 above)
Comment (nmtien) [reply nats/vaz.yml]: DONE

File: nats/subsz.yml (hunk 5 above)
Comment (nmtien) [reply nats/subsz.yml]: I will check the real value again, not sure i check it's value last time.

Comment (nmtien) [reply nats/subsz.yml]: It is counter, i also change some other metric type to counter

File: nats/subsz.yml (hunk 2 above)
Comment (sivasamyk) [reply nats/subsz.yml]: OK. I also see that the actual cache hit metric is not exposed (to calculate the rate ourselves). Please mention that the value range is from 0-1 in the description.

Comment (nmtien) [reply nats/subsz.yml]: OK. Done.

File: nats-server/connz.yml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+type: json
+data:
+  url: /connz
+  server: ${NATS_SERVER_HOSTPORT}
+
+observation:
+  - name: connz
+    metricNamespace: nats_server
```
Comment (sivasamyk) [inline nats-server/connz.yml]: the namespace can be `nats.server`

File: nats-server/connz.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    metric:
+      - name: connections
+        source: num_connections
+        type: long_gauge
+        label: connections
+        description: The number of connections
+
+    tag:
+      - name: nats.servers.server
```
Comment (sivasamyk) [inline nats-server/connz.yml]: tag name can be `nats.server.id`. `nats.servers.server` seems to stutter

Comment (nmtien) [reply nats-server/connz.yml]: OK, DONE.

File: nats-server/connz.yml (hunk 9 above)
Comment (nmtien) [reply nats-server/connz.yml]: DONE.

File: nats/subsz.yml (hunk 2 above)
Comment (nmtien) [reply nats/subsz.yml]: i add value range (0-1) to metric description.

File: nats-server/connz.yml (hunk 10 above)
Comment (sivasamyk) [reply nats-server/connz.yml]: Hi Tien, Bojan and I had a discussion. Shall we change the metric namespace to `nats` and prefix all metric names with `server`? For the streaming server (when we add the integration) we can prefix the metric namespace with `streaming.server`

The metric name will be like `server. connections`, `server.subscriptions.removes` etc

File: nats-server/subsz.yml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        source: num_cache
+        type: long_gauge
+        label: cache size
+        description: The number of result sets in the cache
+
+      - name: server.cache.hit.rate
+        source: cache_hit_rate
+        type: double_gauge
+        label: cache hit rate
+        description: The rate matches are being retrieved from cache (value is in [0-1])
```
Comment (sivasamyk) [inline nats-server/subsz.yml]: enclose this description with-in double quotes

**Comment by otisg** [comment]
@nmtien anything left to do here?

File: nats-server/subsz.yml (hunk 11 above)
Comment (nmtien) [reply nats-server/subsz.yml]: DONE

**Comment by nmtien** [comment]
Should i merge, nothing left from me.

**Pull request merged**